From 78fccb6df8034d0ccd53feb2edfb3e36730527bf Mon Sep 17 00:00:00 2001
From: Alex Porosanu <alexandru.porosanu@freescale.com>
Date: Tue, 4 Aug 2015 23:32:48 +0300
Subject: [PATCH 087/146] fsl_qman: fix CGR congestion indication processing

QM_MCR_QUERYCONGESTION macro checks if the corresponding bit for
the CGR is set in the "query congestion" command response. If
the corresponding bit is set (meaning that the respective CGR
is in a congested state), then the corresponding callback
is called.
The bit-checking must be done by taking into account the difference
between the DPAA block's and the core's endianness.

Note: improper functioning of CGR callback (i.e. not being called
even if congestion is present) leads to packets being accumulated
in one or more FQs which will eventually lead in the end to
exhausting all the system memory (OOM)

Signed-off-by: Alex Porosanu <alexandru.porosanu@freescale.com>
---
 include/linux/fsl_qman.h | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/include/linux/fsl_qman.h b/include/linux/fsl_qman.h
index 2a154eb..1ea11d4 100644
--- a/include/linux/fsl_qman.h
+++ b/include/linux/fsl_qman.h
@@ -1635,7 +1635,8 @@ struct qm_mc_result {
 static inline int QM_MCR_QUERYCONGESTION(struct __qm_mcr_querycongestion *p,
 					u8 cgr)
 {
-	return p->__state[__CGR_WORD(cgr)] & (0x80000000 >> __CGR_SHIFT(cgr));
+	return be32_to_cpu(p->__state[__CGR_WORD(cgr)]) &
+	       (0x80000000 >> __CGR_SHIFT(cgr));
 }
 
 
-- 
2.1.0.27.g96db324

